blog/2020/08/invalidations.md

```diff
@@ -0,0 +1,367 @@
+
+You can see that Julia no longer uses that optimization for when `container[1]` happens to be an `Int`.
+Experience has shown that once a method has a couple of specializations, it may accumulate yet more,
+and to reduce the amount of recompilation needed Julia quickly abandons attempts to optimize the
+case handling containers with unknown element types.
+If you look carefully, you'll notice one remaining optimization: the call to `Main.f` is type-asserted as `Int`,
+since both known methods of `f` return an `Int`.
+This can be a very useful performance optimization for any code that uses the output of `applyf`.
+But define two additional methods of `f`,
+
+```julia
+julia> f(::String) = 3
+f (generic function with 4 methods)
+
+julia> f(::Dict) = 4
+f (generic function with 4 methods)
+
+julia> @code_typed applyf(Any[true])
+CodeInfo(
+1 ─ %1 = Base.arrayref(true, container, 1)::Any
+│   %2 = Main.f(%1)::Any
+└──      return %2
+) => Any
+```
+
+and you'll see that even this optimization is now abandoned.
+This is because checking lots of methods, to see if they all return `Int`, is too costly an operation to be worth doing in all cases.
+
+Altogether, `apply(::Vector{Any})` has been compiled three times: once when `f(::Int)` was the only method for `f`,
+once where there were two methods, and once when there were four.
+Had Julia known in advance that we'd end up here, it never would have bothered to compile those intermediate versions;
+recompilation takes time, and we'd rather avoid spending that time if it's not absolutely necessary.
+Fortunately, from where we are now Julia has stopped making assumptions about how many methods of `f` there will be,
+so from this point forward we won't need to recompile `applyf` for a `Vector{Any}` even if we define further methods of `f`.
+
+Recompilation is triggered by two events:
+- when a new method is defined, old compiled methods that made now-incorrect assumptions get *invalidated*
+- when no valid compiled method can be found to handle a call, Julia (re)compiles the necessary code.
+
+It's the moment of method definition, triggering invalidations of other previously-compiled code,
+that will be the focus of this blog post.
+
+## How common is method invalidation?
+
+Unfortunately, method invalidation is (or rather, used to be) pretty common.
+First, let's get some baseline statistics.
+Using the [MethodAnalysis] package, you can find out that a fresh Julia session has almost 50,000 `MethodInstance`s tucked away in its cache.
+These are mostly for `Base` and the standard libraries.
+(There are some additional `MethodInstance`s that get created to load the MethodAnalysis package and do this analysis, but these are surely a very small fraction of the total.)
+
+Using [SnoopCompile], we can count the number of invalidations triggered by loading various packages into a fresh Julia session:
+
+| Package | Version | # invalidations (Julia 1.5) | # invalidations (Julia 1.6) |
```

Comment:
Suggestion.. Perhaps it's better to reference this as `master (githash, 1.6-pre)`, given that 1.6 is fast evolving? I'm sure this post will be well-visited beyond the finalization of 1.6, so just saying 1.6 might confuse things a little?

Reply:
That was a good suggestion. Enough of my PRs have been merged that I can now comfortably commit to a git-sha, so let's merge this sucker.